File: simply_lifestyle_app/lib/domain/models/order/order.dart
import 'package:simply_lifestyle_app/domain/models/entity.dart';
import 'package:simply_lifestyle_app/domain/models/order/order_item.dart';
import 'package:simply_lifestyle_app/domain/models/order/order_status.dart';

class Order extends Entity {
  Order(
      {super.id,
      this.customerId,
      this.customerFirstName,
      this.customerLastName,
      required this.orderDate,
      required this.deliveryDate,
      required this.status,
      required this.orderItems});

  final String? customerId;
  final String? customerFirstName;
  final String? customerLastName;
  final DateTime orderDate;
  final DateTime deliveryDate;
  final OrderStatus status;
  final List<OrderItem> orderItems;

  factory Order.fromJson(Map<String, dynamic> json) {
    return switch (json) {
      {
        'id': String id,
        'customerId': String customerId,
        'customerFirstName': String customerFirstName,
        'customerLastName': String customerLastName,
        'orderDate': DateTime orderDate,
        'deliveryDate': DateTime deliveryDate,
        'status': OrderStatus status,
        'orderItems': List<Map<String, dynamic>> orderItems
      } =>
        Order(
            id: id,
            customerId: customerId,
            customerFirstName: customerFirstName,
            customerLastName: customerLastName,
            orderDate: orderDate,
            deliveryDate: deliveryDate,
            status: status,
            orderItems: orderItems.map((o) => OrderItem.fromJson(o)).toList()),
      _ => throw const FormatException('Failed to load order.'),
    };
  }
}

File: SimplyLifestyleApi/OrderManagement/OrderManagement.Domain/Models/Orders/Order.cs
using Common.Domain;

namespace OrderManagement.Domain;

public class Order : Entity, IAggregateRoot
{
    public HashSet<OrderItem> OrderItems { get; private set; }

    public Order(Guid customerId, DateTimeOffset deliveryDate)
    {
        ValidateDeliveryDate(deliveryDate);

        CustomerId = customerId;
        OrderDate = DateTimeOffset.UtcNow;
        OrderItems = new HashSet<OrderItem>();
        Status = OrderStatus.Pending;
        
        RaiseEvent(new OrderAddedEvent());
    }

    public Guid CustomerId { get; private set; }
    public DateTimeOffset OrderDate { get; private set; }
    public DateTimeOffset DeliveryDate { get; private set; }
    public OrderStatus Status { get; private set; }

    public Order AddOrderItem(Guid productId, int quantity)
    {
        ValidateOrderItemQuantity(quantity);

        var orderItem = new OrderItem(Id, productId, quantity);
        OrderItems.Add(orderItem);

        return this;
    }

    public Order RemoveOrderItem(Guid orderItemId)
    {
        var orderItem = OrderItems.FirstOrDefault(oi => oi.Id == orderItemId);
        if (orderItem != null)
        {
            OrderItems.Remove(orderItem);
        }
        return this;
    }

    public Order UpdateStatus(OrderStatus status)
    {
        Status = status;
        return this;
    }

    public Order UpdateCustomerId(Guid customerId)
    {
        CustomerId = customerId;
        return this;
    }

    public Order UpdateOrderDate(DateTimeOffset orderDate)
    {
        OrderDate = orderDate;
        return this;
    }

    public Order UpdateDeliveryDate(DateTimeOffset deliveryDate)
    {
        ValidateDeliveryDate(deliveryDate);
        DeliveryDate = deliveryDate;
        return this;
    }

    public Order UpdateOrderItem(Guid orderItemId, Guid productId, int quantity)
    {
        ValidateOrderItemQuantity(quantity);

        var orderItem = OrderItems.FirstOrDefault(oi => oi.Id == orderItemId);
        if (orderItem != null)
        {
            orderItem.UpdateProductId(productId);
            orderItem.UpdateQuantity(quantity);
        }
        return this;
    }

    private void ValidateDeliveryDate(DateTimeOffset deliveryDate)
    {
        if (deliveryDate < DateTime.UtcNow)
        {
            throw new ArgumentException("Order date cannot be in the future.");
        }
    }

    private void ValidateOrderItemQuantity(int quantity)
    {
        if (quantity < OrderModelConstants.OrderItem.MinQuantity || quantity > OrderModelConstants.OrderItem.MaxQuantity)
        {
            throw new ArgumentException($"Order item quantity must be between {OrderModelConstants.OrderItem.MinQuantity} and {OrderModelConstants.OrderItem.MaxQuantity}.");
        }
    }
}
